Dockerfile

```diff
@@ -40,9 +41,15 @@ COPY . .
 
 # Show build caching stats to check if it was used in the end.
 # Has to be the part of the same RUN since cachepot daemon is killed in the end of this RUN, losing the compilation stats.
-RUN set -e \
-&& mold -run cargo build --locked --release \
-    && cachepot -s
+RUN --mount=type=cache,uid=1000,target=/usr/local/cargo/registry \
```

Comment:
I think, we should cache both directories, similar way we do it now with "native" rust builds:

https://github.com/neondatabase/neon/blob/a463749f59c3fe020065c4cacc91df8fa11ffb99/.github/workflows/build_and_test.yml#L113-L125

The way it's defined now, we're mixing text source files (` ~/.cargo/registry/ `) and the `target/` binaries, the latter are platform-specific, so if we copy-paste the whole thing, the cache key should be different.

Reply:
I added `/usr/local/cargo/git/target` to cache, if I understood you right.

Reply:
Sorry for cryptic comment, my main concern are the `/home/nonroot/target/release/` files: those are supposed to be cached between agents somehow? So that some other Docker-in-Docker builds reuses the `target/` to build incrementally on top only things that were changed?

In the link I've shared above, we archive and upload that `target/` (among others) and then download it before building things and I would expect for similar things to happen here.

Reply:
The original comment in slack was about local builds, so was aiming to speed up only rebuilds on the local machines. These cache mounts use local machine cache, so they speed up only when:

1. docker build
2. edit local rust file
3. docker build

And on GH agents my changes are unlikely to introduce any significant changes. What kind of agents do we use? Standard GHA agents? Or self-hosted runners?

I also added [sharing=locked](https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/reference.md#run---mounttypecache) to reduce changes to concurrent access to cache.

Reply:
> self-hosted runners

These.

> about local builds

Now I start to understand, never mind me then.
In that case, before we make our mind on this for CI (yet if this approach works on our build agents, I think it's worth trying it), never mind my words about `cachepot`, we should not remove it.
It could be disabled for local builds with `RUSTC_WRAPPER=""` in case it helps more.